### PR TITLE
Converted wysiwyg jasmine file to jest test fixes #784

### LINKED ISSUE #784: Create and port wysiwyg jasmine file to jest test

### PR DESCRIPTION
**Pull request by NARUDESIGNS**
Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `grunt jasmine`
* [x] code is in uniquely-named feature branch and has no merge conflicts
* [x] PR is descriptively titled
* [x] PR body includes `fixes #0000`-style reference to original issue #
* [x] ask `@publiclab/reviewers` for help, in a comment below

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software

We have a loose schedule of reviewing and pulling in changes every Tuesday and Friday, and publishing changes on Fridays. 

Thanks!

### CODE REVIEW

**Comment by jywarren** [comment]
Hi, it looks like, since we recently merged in some new changes to the main branch, you need to rebase your work on top of the latest main. You can do that by following the process outlined here, but you may be able to skip step 2 if your main is still in sync with publiclab's main: https://publiclab.org/wiki/contributing-to-public-lab-software#Rewinding+the+master+branch

In general there are more tips and guides on rebasing here: https://publiclab.org/wiki/developers#Resources

Thanks!

**Comment by jywarren** [comment]
Looks like you're seeing real test failures now, not just an error before running. Let me know if you need help with these!

**Comment by NARUDESIGNS** [comment]
I ran the test on my local machine and it all passed. How can I run it here please?
Or can you tell me the error it throws in the log?

**Comment by jywarren** [comment]
Sure - it's here: https://github.com/publiclab/PublicLab.Editor/runs/4778470519?check_suite_focus=true#step:7:49

What i do is click the Details link, then click the line in the log with the red error icon and it expands, then i search for "FAIL" to find individual failures. The log is tough to read so the search function really helps!

image

**Comment by jywarren** [comment]
One looks like a timeout, which is a little unexpected. Let's try to figure out why the others are true/false failures, and maybe the timeout will resolve itself? We can also just re-run it to see if it was something intermittent if it really is supposed to run smoothly. Finally, we could lengthen the timeout limit to see if it helps! But, my guess is it should be enough time, we're just waiting for local asynchronous code to run, which should be pretty fast.

**Comment by NARUDESIGNS** [comment]
This is a bit confusing. Line 49 refers to `bold.test.js`, all these tests passed before I started learning about them. I wonder how it is failing now when no changes was made to it.
@jywarren

**Comment by jywarren** [comment]
Ah, ok. So these test failures are totally unconnected to this PR. That's usually a signal we should a) restart them and b) look at other unrelated reasons the tests could be flaky, but in a separate issue. Would you mind opening one and linking specifically to this run, so we can check the logs? This URL shouldn't change even if we start the tests over:

https://github.com/publiclab/PublicLab.Editor/runs/4778470519

And you can link to this PR as an example of one where this happened.

**Comment by jywarren** [comment]
Yes, so confirming, this does seem to be intermittent, and unfortunately this time the CustomInsertTest failed:

```
FAIL test/ui-testing/CustomInsert.test.js (10.573 s)
  Custom Insert text
    ✕ Add Custom Insert text in rich text mode (10002 ms)

  ● Custom Insert text › Add Custom Insert text in rich text mode

    : Timeout - Async callback was not invoked within the 10000 ms timeout specified by jest.setTimeout.Timeout - Async callback was not invoked within the 10000 ms timeout specified by jest.setTimeout.Error:

       7 | 
       8 | describe('Custom Insert text', () => {
    >  9 |   test('Add Custom Insert text in rich text mode', async () => {
         |   ^
```

My guess is we may need to restart it again, but also copy these errors into that new tracking issue and try to work the problem there. I'll also try running it in GitPod to see!

File: test/ui-testing/wysiwyg.test.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
+const timeout = process.env.SLOWMO ? 60000 : 10000;
```
Comment (jywarren) [inline test/ui-testing/wysiwyg.test.js:1]: We can try increasing this to above 10 seconds for all tests...?

Comment (NARUDESIGNS) [reply test/ui-testing/wysiwyg.test.js:1]: Sure we could do that. But may I ask please, why do you think we should increase the timeout? Did you experience any timeout recently?
@jywarren

Comment (jywarren) [reply test/ui-testing/wysiwyg.test.js:1]: Based on the message in the logs:

`    : Timeout - Async callback was not invoked within the 10000 ms timeout specified by jest.setTimeout.Timeout - Async callback was not invoked within the 10000 ms timeout specified by jest.setTimeout.Error:`

I am guessing that for some reason it took longer. It's worth a try!

Comment (jywarren) [reply test/ui-testing/wysiwyg.test.js:1]: And that was the same as with one of the earlier bold test failures!

Comment (NARUDESIGNS) [reply test/ui-testing/wysiwyg.test.js:1]: Oh ok, I get. I think I got that error sometimes and when I ran the test again it works fine. But it's ok, we could increase the timeout value to say 15000ms or 20000ms. 
Is that much or enough, what do you think?

**Comment by jywarren** [comment]
I think if we increase it to 15 or 20 seconds and it doesn't help, then we
at least know that our problem has to be solved another way, and it's not
like we're just a second or two off. 5 or 10 seconds on each of... 6 tests?
Isn't too bad, it's 30-60 seconds overall.

**Comment by NARUDESIGNS** [comment]
Alright. Lastly, can we try running that particular test again here? 
Can you please teach me how to do it so I can try running again it to see if I'd always get that error? (just curious)

**Comment by jywarren** [comment]
For sure. Either you close and reopen the PR, or you can open the checks and you'll see a button:

image

Where it says `Cancel workflow` it'll later say "re-run" (once it's done). You can do it there too. You could cancel too but i'm always surprised by how long it takes to cancel.

**Comment by jywarren** [comment]
OK, i just pushed a commit lengthening all Jest test timeouts to 15 seconds instead of 10. Let's see... 🤞

**Comment by jywarren** [comment]
Well, 15 seconds doesn't seem to have worked :-(

This is weird because it just started happening, right? A variety of earlier PRs didn't see this issue?

I'm going to check if somehow there's an error happening inside the async test and not getting logged out. Maybe it's not a timeout but an error somehow...

**Comment by jywarren** [comment]
I think our next step is to research what's causing this timeout error. I started searching in https://github.com/facebook/jest/search?q=Timeout+-+Async+callback+was+not+invoked+within&type=issues - and bearing in mind https://jestjs.io/docs/asynchronous

I can try again later in the week!

**Comment by NARUDESIGNS** [comment]
> Well, 15 seconds doesn't seem to have worked :-(
> 
> This is weird because it just started happening, right? A variety of earlier PRs didn't see this issue?
> 
> I'm going to check if somehow there's an error happening inside the async test and not getting logged out. Maybe it's not a timeout but an error somehow...

The PRs I created didn't have this issues and also my local tests passes as well. Where are you seeing these? I'm a little confused 🤕

**Comment by jywarren** [comment]
Just copying in a note here:

> I think we need to read through some issues in the Jest repository, searching for keywords to find something similar to what we're seeing. We may also want to search a bit in Stack Overflow. Just trying to find someone who's seen something similar! We could also try to run things more verbosely so we get more log output as well... to try to determine when it's failing.

**Comment by jywarren** [comment]
Ideas

- tests in parallel - adding a test somehow overloads the system? Or, this test is erroring, and the browser is then hung up when it's needed by the other tests?
- make test runs 25 seconds long (not ideal but just in case)
- Add a timeout to beforeAll - https://stackoverflow.com/questions/51048119/jest-async-callback-was-not-invoked-within-the-5000ms-timeout-specified-by-jes
- start a completely new PR with this test to avoid caching... maybe superstitious :-)
- catch errors within the async test: https://github.com/puppeteer/puppeteer/issues/3870#issuecomment-459459577

**Comment by NARUDESIGNS** [comment]
Hey @jywarren I think the solution is solid and we got all our answers. I just pasted the changes from #804, reduced the timeout from 60000 to 10000, added comment to explain that part and pushed here which was the original PR and it passes the test. 
I think we solved it!

**Review by jywarren — APPROVED** [review]

**Comment by jywarren** [comment]
Super well done! This was a tough and mysterious one but you did great. 🎉

**Pull request merged**